netcore/System.Private.CoreLib/src/System/Environment.Unix.cs

```diff
@@ -13,36 +13,91 @@ namespace System
 {
 	partial class Environment
 	{
+		private static Dictionary<string, string> s_environment;
+
 		static string GetEnvironmentVariableCore (string variable)
 		{
-			using (var h = RuntimeMarshal.MarshalString (variable)) {
-				return internalGetEnvironmentVariable_native (h.Value);
+			Debug.Assert(variable != null);
+
+			if (s_environment == null) {
+				using (var h = RuntimeMarshal.MarshalString (variable)) {
+					return internalGetEnvironmentVariable_native (h.Value);
+				}
+			}
+
+			lock (s_environment) {
+				variable = TrimStringOnFirstZero (variable);
+				s_environment.TryGetValue (variable, out string value);
+				return value;
 			}
 		}
 
-		[MethodImplAttribute (MethodImplOptions.InternalCall)]
-		extern static string internalGetEnvironmentVariable_native (IntPtr variable);
+		static unsafe void SetEnvironmentVariableCore (string variable, string value)
+		{
+			Debug.Assert(variable != null);
 
-		[MethodImplAttribute (MethodImplOptions.InternalCall)]
-		private extern static string [] GetEnvironmentVariableNames ();
+			EnsureEnvironmentCached ();
+			lock (s_environment) {
+				variable = TrimStringOnFirstZero (variable);
+				value = value == null ? null : TrimStringOnFirstZero (value);
+				if (string.IsNullOrEmpty (value)) {
+					s_environment.Remove (variable);
+				} else {
+					s_environment[variable] = value;
```

Comment:
Could be just `s_environment [variable] = TrimStringOnFirstZero (value);`

Reply:
No, the trimming could make the value empty and unset the environment variable.